hermes/probe/delete_file.go

```diff
@@ -0,0 +1,104 @@
+//	- client: pass an initialised storage client for this target system.
+//	- logger: pass the logger associated with the probe calling this function.
+// Returns:
+//	- fileID: returns the ID of the file delete OR a missing file to be created if one is found.
+//	- err:
+//		- [...].hermes_file_missing: the file to be deleted does not exist in Hermes' StateJournal.
+//		- [...].file_missing: the file to be deleted could not be found in the target bucket.
+//		- [...].bucket_missing: the target bucket on this target system was not found.
+//		- [...].probe_failed: there was an error during one of the API calls and the probe failed.
+//		- [...].deleted_file_found: the file was deleted but it is still found in the target bucket.
+//		- [...].list_bucket_failed: the listBucket operation failed when checking if the target file was deleted.
+func DeleteRandomFile(ctx context.Context, config *probepb.HermesProbeDef, target *Target, client *stiface.Client, logger *logger.Logger) (int, error) {
+	bucket := target.Target.GetBucketName()
+
+	fileID := hermes.PickFileToDelete()
+
+	filename, ok := target.Journal.Filenames[int32(fileID)]
+	if !ok {
+		return fileID, fmt.Errorf("deleteRandomFile(id: %d).hermes_file_missing: could not delete file %s, file not found", fileID, filename)
+	}
+
+	file := (*client).Bucket(bucket).Object(filename)
+
+	start := time.Now()
+	if err := file.Delete(ctx); err != nil {
+		var status string
+		if err == storage.ErrObjectNotExist {
+			status = m.ExitStatuses[m.FILE_MISSING]
+		} else if err == storage.ErrBucketNotExist {
+			status = m.ExitStatuses[m.BUCKET_MISSING]
+		} else {
+			status = m.ExitStatuses[m.PROBE_FAILED]
+		}
+		target.LatencyMetrics.ApiCallLatency[m.ApiCalls[m.API_DELETE_FILE]][status].Metric("latency").AddFloat64(time.Now().Sub(start).Seconds())
+		return fileID, fmt.Errorf("deleteRandomFile(id: %d).%s: could not delete file %s: %w", fileID, status, filename, err)
+	}
+	target.LatencyMetrics.ApiCallLatency[m.ApiCalls[m.API_DELETE_FILE]][m.ExitStatuses[m.SUCCESS]].Metric("latency").AddFloat64(time.Now().Sub(start).Seconds())
+
+	query := &storage.Query{Prefix: filename}
+	start = time.Now()
+	it := (*client).Bucket(bucket).Objects(ctx, query)
+	target.LatencyMetrics.ApiCallLatency[m.ApiCalls[m.API_LIST_FILES]][m.ExitStatuses[m.SUCCESS]].Metric("latency").AddFloat64(time.Now().Sub(start).Seconds())
+	for {
+		obj, err := it.Next()
+		if obj.Name == filename {
+			return fileID, fmt.Errorf("deleteRandomFile(id %d).deleted_file_found: object %v in bucket %q still listed after delete", fileID, *obj, bucket)
+		}
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return fileID, fmt.Errorf("deleteRandomFile(id: %d).list_bucket_failed: unable to list bucket %q: %w", fileID, bucket, err)
+		}
+	}
+
+	// Update in-memory NIL file after delete operation.
+	delete(target.Journal.Filenames, int32(fileID))
+
+	logger.Infof("Object %v deleted in bucket %s.", file, bucket)
+	return fileID, nil
```

Comment:
What's the purpose of returning the file ID?

Reply:
The ID is returned as it is used in the next function in the algorithm, createFile(). Rather than having to search for the one missing file ID in createFile(), I thought it would be better to return it and pass it to createFile().